app/questionnaire/questionnaire_manager.py

```diff
@@ -178,130 +50,79 @@ def validate(self, post_data):
-
-                    # do any routing
-                    next_location = self.navigator.get_next_location(self.get_answers(), location)
-                    logger.info("next location after routing is %s", next_location)
-
-                    # go to that location
-                    self.go_to_node(next_location)
-                    logger.debug("Going to location %s", next_location)
-                except UserActionProcessorException as e:
-                    logger.error("Error processing user actions")
-                    logger.exception(e)
-            else:
-                # bug fix for using back button which then fails validation
-                self.go_to_node(self.get_current_location())
 
-            # now return the location
-            current_location = self.get_current_location()
-            logger.debug("Returning location %s", current_location)
-            return is_valid
-        else:
-            raise InvalidLocationException()
+        # process incoming post data
+        user_action = self._get_user_action(post_data)
+
+        is_valid = self.validate(location, post_data)
+        # run the validator to update the validation_store
+        if is_valid:
+
+            # Store answers in QuestionnaireStore
+            questionnaire_store = get_questionnaire_store(current_user.user_id, current_user.user_ik)
+
+            if location not in questionnaire_store.visited_blocks:
+                questionnaire_store.visited_blocks.append(location)
+
+            questionnaire_store.save()
+
+            # process the user action
+            try:
+                user_action_processor = UserActionProcessor(self._schema, self)
+                user_action_processor.process_action(user_action)
+
+            except UserActionProcessorException as e:
+                logger.error("Error processing user actions")
+                logger.exception(e)
+                return False, e.next_location
+
+        return is_valid, None
 
     def get_rendering_context(self, location, is_valid=True):
-        node = self._current
-        state_items = []
+
         if is_valid:
             if location == 'summary':
                 return self.get_summary_rendering_context()
             else:
-                self.build_state(node, node.answers)
+                # apply page answers?
+                self.build_state(location, get_answers(current_user))
 
         if self.state:
             self._plumbing_preprocessing(self.state)
             self._conditional_display(self.state)
 
-        if self.state not in state_items:
-            state_items.append(self.state)
+        previous_location = self.navigator.get_previous_location(get_answers(current_user), location)
 
         # look up the preprocessor and then build the view data
-        return build_questionnaire_model(self._json, state_items)
+        return build_questionnaire_model(self._json, [self.state], previous_location)
```

Comment:
Yes i can now. i left it like that as it was being used for summary, but now isn't so ill update that.